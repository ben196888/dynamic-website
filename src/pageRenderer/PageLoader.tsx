import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { State } from '../store/Store';
import {
  ElementState, Element, TreeState,
} from '../store/reducers/types';
import { elements } from '../mockPage';
import factory from './RenderFactory';
import { setElements } from '../store/reducers/actions';

export default function PageLoader() {
  const [loading, setLoading] = useState(0);
  const dispatch = useDispatch();
  const tree = useSelector<State, TreeState>((state) => state.tree);
  const nodeMap = useSelector<State, ElementState>((state) => state.elements);
  useEffect(() => {
    Promise.resolve(elements)
      .then((result: Element[]) => {
        dispatch(setElements(result));
        setLoading(2);
      })
      .catch(() => {
        setLoading(1);
      });
  }, [dispatch]);
  let text: string = '';
  if (loading === 0) { text = 'Now loading'; }
  if (loading === 1) { text = 'Load failed'; }
  return (
    <>
      {text}
      {tree.map((e) => factory(e, nodeMap))}
    </>
  );
}
