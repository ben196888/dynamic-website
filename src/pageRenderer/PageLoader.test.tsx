import React from 'react';
import { render, act } from '../testUtils';
import PageLoader from './PageLoader';

describe('<PageLoader />', () => {
  test('should dispatch set element action', async () => {
    const { container } = render(<PageLoader />, {
      initialState: { tree: [], elements: {} },
    });
    await act(async () => {
      render(<PageLoader />);
    });
    expect(container).toMatchSnapshot();
  });
});
