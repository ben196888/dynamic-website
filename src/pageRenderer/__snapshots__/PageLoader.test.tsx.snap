// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageLoader /> should dispatch set element action 1`] = `
<div>
  
  <div
    type="div"
  >
    Hello world 
    <span
      type="span"
    >
      John Doe
    </span>
  </div>
  <a
    href="https://google.com"
    type="a"
  >
    link
  </a>
  <br
    type="br"
  />
  <button
    type="button"
  >
    click to alert
  </button>
  Hello World
  <br
    type="br"
  />
  String can render straight away
</div>
`;
